src/workerd/server/channel-token.c++

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) 2025 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "channel-token.h"
+
+#include <workerd/server/channel-token.capnp.h>
+#include <workerd/util/entropy.h>
+
+#include <openssl/evp.h>
+#include <openssl/sha.h>
+
+#include <capnp/serialize-packed.h>
+#include <kj/io.h>
+
+// It's 2025, nobody uses big-endian anymore. But just in case someone tries, flag it here.
+// Specifically, the magic number in TokenHeader is encoded in host order.
+#if !defined(__LITTLE_ENDIAN__) || __BYTE_ORDER != __LITTLE_ENDIAN
+#error "This code assumes little-endian architecture."
+#endif
+
+namespace workerd::server {
+
+ChannelTokenHandler::ChannelTokenHandler(Resolver& resolver): resolver(resolver) {
+  getEntropy(tokenKey);
+
+  SHA256_CTX ctx{};
+  KJ_ASSERT(SHA256_Init(&ctx));
+  KJ_ASSERT(SHA256_Update(&ctx, tokenKey, sizeof(tokenKey)));
+
+  byte hash[SHA256_DIGEST_LENGTH]{};
+  KJ_ASSERT(SHA256_Final(hash, &ctx));
+
+  static_assert(KEY_ID_SIZE <= SHA256_DIGEST_LENGTH);
+  kj::arrayPtr(keyId).copyFrom(kj::arrayPtr(hash).first(KEY_ID_SIZE));
+}
+
+kj::Array<byte> ChannelTokenHandler::encodeChannelTokenImpl(ChannelToken::Type type,
+    IoChannelFactory::ChannelTokenUsage usage,
+    kj::StringPtr serviceName,
+    kj::Maybe<kj::StringPtr> entrypoint,
+    Frankenvalue& props) {
+  capnp::word scratch[128]{};
+  capnp::MallocMessageBuilder message(scratch);
+  auto builder = message.getRoot<ChannelToken>();
+
+  builder.setType(type);
+
+  // If requireAllowsTransfer() passed, then we are not dynamic so should have a service name.
```

Comment:
Vestigial comment? It looks like the `requireAllowsTransfer()` happens at call sites.